File: korge-ext-spriter/src/com/brashmonkey/spriter/Entity.java
package com.brashmonkey.spriter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 * Represents an entity of a Spriter SCML file.
 * An entity holds {@link Animation}s, an {@link #id}, a {@link #name}.
 * {@link #characterMaps} and {@link #objectInfos} may be empty.
 * @author Trixt0r
 *
 */
public class Entity {

    public final int id;
    public final String name;
    private final Animation[] animations;
    private int animationPointer = 0;
    private final HashMap<String, Animation> namedAnimations;
    private final CharacterMap[] characterMaps;
    private int charMapPointer = 0;
    private final ObjectInfo[] objectInfos;
    private int objInfoPointer = 0;
	
	Entity(int id, String name, int animations, int characterMaps, int objectInfos){
		this.id = id;
		this.name = name;
		this.animations = new Animation[animations];
		this.characterMaps = new CharacterMap[characterMaps];
		this.objectInfos = new ObjectInfo[objectInfos];
		this.namedAnimations = new HashMap<String, Animation>();
	}
	
	void addAnimation(Animation anim){
		this.animations[animationPointer++] = anim;
		this.namedAnimations.put(anim.name, anim);
	}
	
	/**
	 * Returns an {@link Animation} with the given index.
	 * @param index the index of the animation
	 * @return the animation with the given index
	 * @throws IndexOutOfBoundsException if the index is out of range
	 */
	public Animation getAnimation(int index){
		return this.animations[index];
	}
	
	/**
	 * Returns an {@link Animation} with the given name.
	 * @param name the name of the animation
	 * @return the animation with the given name or null if no animation exists with the given name
	 */
	public Animation getAnimation(String name){
		return this.namedAnimations.get(name);
	}
	
	/**
	 * Returns the number of animations this entity holds.
	 * @return the number of animations
	 */
	public int animations(){
		return this.animations.length;
	}
	
	/**
	 * Returns whether this entity contains the given animation.
	 * @param anim the animation to check
	 * @return true if the given animation is in this entity, false otherwise.
	 */
	public boolean containsAnimation(Animation anim){
		for(Animation a: this.animations)
			if(a == anim) return true;
		return false;
	}
	
	/**
	 * Returns the animation with the most number of time lines in this entity.
	 * @return animation with the maximum amount of time lines.
	 */
	public Animation getAnimationWithMostTimelines(){
		Animation maxAnim = getAnimation(0);
		for(Animation anim: this.animations){
			if(maxAnim.timelines() < anim.timelines()) maxAnim = anim;
		}
		return maxAnim;
	}
    
    /**
     * Returns a {@link CharacterMap} with the given name.
     * @param name name of the character map
     * @return the character map or null if no character map exists with the given name
     */
    public CharacterMap getCharacterMap(String name){
    	for(CharacterMap map: this.characterMaps)
    		if(map.name.equals(name)) return map;
    	return null;
    }
    
    void addCharacterMap(CharacterMap map){
    	this.characterMaps[charMapPointer++] = map;
    }
    
    void addInfo(ObjectInfo info){
    	this.objectInfos[objInfoPointer++] = info;
    }
    
    /**
     * Returns an {@link ObjectInfo} with the given index.
     * @param index the index of the object info
     * @return the object info
     * @throws IndexOutOfBoundsException if index is out of range
     */
    public ObjectInfo getInfo(int index){
    	return this.objectInfos[index];
    }
    
    /**
     * Returns an {@link ObjectInfo} with the given name.
     * @param name name of the object info
     * @return object info or null if no object info exists with the given name
     */
    public ObjectInfo getInfo(String name){
    	for(ObjectInfo info: this.objectInfos)
    		if(info.name.equals(name)) return info;
    	return null;
    }
    
    /**
     * Returns an {@link ObjectInfo} with the given name and the given {@link ObjectType} type.
     * @param name the name of the object info
     * @param type the type if the object info
     * @return the object info or null if no object info exists with the given name and type
     */
    public ObjectInfo getInfo(String name, ObjectType type){
    	ObjectInfo info = this.getInfo(name);
    	if(info != null && info.type == type) return info;
    	else return null;
    }
    
    /**
     * Represents the object types Spriter supports.
     * @author Trixt0r
     *
     */
    public static enum ObjectType{
    	Sprite, Bone, Box, Point, Skin;
    	
    	/**
    	 * Returns the object type for the given name
    	 * @param name the name of the type
    	 * @return the object type, Sprite is the default value
    	 */
    	public static ObjectType getObjectInfoFor(String name){
    		if(name.equals("bone")) return Bone;
    		else if(name.equals("skin")) return Skin;
    		else if(name.equals("box")) return Box;
    		else if(name.equals("point")) return Point;
    		else return Sprite;
    	}
    }
    
    /**
     * Represents the object info in a Spriter SCML file.
     * An object info holds a {@link #type} and a {@link #name}.
     * If the type is a Sprite it holds a list of frames. Otherwise it has a {@link #size} for debug drawing purposes.
     * @author Trixt0r
     *
     */
    public static class ObjectInfo{
    	public final ObjectType type;
    	public final List<FileReference> frames;
    	public final String name;
    	public final Dimension size;
    	
    	ObjectInfo(String name, ObjectType type, Dimension size, List<FileReference> frames){
    		this.type = type;
    		this.frames = frames;
    		this.name = name;
    		this.size = size;
    	}
    	
    	ObjectInfo(String name, ObjectType type, Dimension size){
    		this(name, type, size, new ArrayList<FileReference>());
    	}
    	
    	ObjectInfo(String name, ObjectType type, List<FileReference> frames){
    		this(name, type, new Dimension(0,0), frames);
    	}
    	
    	public String toString(){
    		return name + ": "+ type + ", size: "+size+"|frames:\n"+frames;
    	}
    }
    
    /**
     * Represents a Spriter SCML character map.
     * A character map maps {@link FileReference}s to {@link FileReference}s.
     * It holds an {@link CharacterMap#id} and a {@link CharacterMap#name}.
     * @author Trixt0r
     *
     */
    public static class CharacterMap extends HashMap<FileReference, FileReference>{
    	private static final long serialVersionUID = 6062776450159802283L;
    	
    	public final int id;
    	public final String name;
    	
    	public CharacterMap(int id, String name){
    		this.id = id;
    		this.name = name;
    	}
    	
    	/**
    	 * Returns the mapped reference for the given key.
    	 * @param key the key of the reference
    	 * @return The mapped reference if the key is in this map, otherwise the given key itself is returned.
    	 */
    	public FileReference get(FileReference key){
    		if(!super.containsKey(key)) return key;
    		else return super.get(key);
    	}
    }
    
    public String toString(){
    	String toReturn = getClass().getSimpleName()+"|[id: "+id+", name: "+name+"]";
    	toReturn +="Object infos:\n";
    	for(ObjectInfo info: this.objectInfos)
    		toReturn += "\n"+info;
    	toReturn +="Character maps:\n";
    	for(CharacterMap map: this.characterMaps)
    		toReturn += "\n"+map;
    	toReturn +="Animations:\n";
    	for(Animation animaton: this.animations)
    		toReturn += "\n"+animaton;
    	toReturn+="]";
    	return toReturn;
    }

}

File: korge-ext-spriter-common/src/main/kotlin/com/soywiz/korge/ext/spriter/com/brashmonkey/spriter/Entity.kt
package com.soywiz.korge.ext.spriter.com.brashmonkey.spriter

import com.soywiz.korio.lang.JvmOverloads

/**
 * Represents an entity of a Spriter SCML file.
 * An entity holds [Animation]s, an [.id], a [.name].
 * [.characterMaps] and [.objectInfos] may be empty.
 * @author Trixt0r
 */
class Entity internal constructor(val id: Int, val name: String, animations: Int, characterMaps: Int, objectInfos: Int) {
	private val animations: Array<Animation>
	private var animationPointer = 0
	private val namedAnimations: HashMap<String, Animation>
	private val characterMaps: Array<CharacterMap>
	private var charMapPointer = 0
	private val objectInfos: Array<ObjectInfo>
	private var objInfoPointer = 0

	init {
		this.animations = Array<Animation>(animations) { Animation.DUMMY }
		this.characterMaps = Array<CharacterMap>(characterMaps) { CharacterMap.DUMMY }
		this.objectInfos = Array<ObjectInfo>(objectInfos) { ObjectInfo.DUMMY }
		this.namedAnimations = HashMap<String, Animation>()
	}

	internal fun addAnimation(anim: Animation) {
		this.animations[animationPointer++] = anim
		this.namedAnimations.put(anim.name, anim)
	}

	/**
	 * Returns an [Animation] with the given index.
	 * @param index the index of the animation
	 * *
	 * @return the animation with the given index
	 * *
	 * @throws IndexOutOfBoundsException if the index is out of range
	 */
	fun getAnimation(index: Int): Animation {
		return this.animations[index]
	}

	/**
	 * Returns an [Animation] with the given name.
	 * @param name the name of the animation
	 * *
	 * @return the animation with the given name or null if no animation exists with the given name
	 */
	fun getAnimation(name: String): Animation? {
		return this.namedAnimations[name]
	}

	/**
	 * Returns the number of animations this entity holds.
	 * @return the number of animations
	 */
	fun animations(): Int {
		return this.animations.size
	}

	/**
	 * Returns whether this entity contains the given animation.
	 * @param anim the animation to check
	 * *
	 * @return true if the given animation is in this entity, false otherwise.
	 */
	fun containsAnimation(anim: Animation): Boolean {
		for (a in this.animations)
			if (a === anim) return true
		return false
	}

	/**
	 * Returns the animation with the most number of time lines in this entity.
	 * @return animation with the maximum amount of time lines.
	 */
	val animationWithMostTimelines: Animation
		get() {
			var maxAnim = getAnimation(0)
			for (anim in this.animations) {
				if (maxAnim.timelines() < anim.timelines()) maxAnim = anim
			}
			return maxAnim
		}

	/**
	 * Returns a [CharacterMap] with the given name.
	 * @param name name of the character map
	 * *
	 * @return the character map or null if no character map exists with the given name
	 */
	fun getCharacterMap(name: String): CharacterMap? {
		for (map in this.characterMaps)
			if (map.name == name) return map
		return null
	}

	internal fun addCharacterMap(map: CharacterMap) {
		this.characterMaps[charMapPointer++] = map
	}

	internal fun addInfo(info: ObjectInfo) {
		this.objectInfos[objInfoPointer++] = info
	}

	/**
	 * Returns an [ObjectInfo] with the given index.
	 * @param index the index of the object info
	 * *
	 * @return the object info
	 * *
	 * @throws IndexOutOfBoundsException if index is out of range
	 */
	fun getInfo(index: Int): ObjectInfo {
		return this.objectInfos[index]
	}

	/**
	 * Returns an [ObjectInfo] with the given name.
	 * @param name name of the object info
	 * *
	 * @return object info or null if no object info exists with the given name
	 */
	fun getInfo(name: String): ObjectInfo? {
		for (info in this.objectInfos)
			if (info.name == name) return info
		return null
	}

	/**
	 * Returns an [ObjectInfo] with the given name and the given [ObjectType] type.
	 * @param name the name of the object info
	 * *
	 * @param type the type if the object info
	 * *
	 * @return the object info or null if no object info exists with the given name and type
	 */
	fun getInfo(name: String, type: ObjectType): ObjectInfo? {
		val info = this.getInfo(name)
		if (info != null && info.type == type)
			return info
		else
			return null
	}

	/**
	 * Represents the object types Spriter supports.
	 * @author Trixt0r
	 */
	enum class ObjectType {
		Sprite, Bone, Box, Point, Skin;


		companion object {

			/**
			 * Returns the object type for the given name
			 * @param name the name of the type
			 * *
			 * @return the object type, Sprite is the default value
			 */
			fun getObjectInfoFor(name: String): ObjectType {
				if (name == "bone")
					return Bone
				else if (name == "skin")
					return Skin
				else if (name == "box")
					return Box
				else if (name == "point")
					return Point
				else
					return Sprite
			}
		}
	}

	/**
	 * Represents the object info in a Spriter SCML file.
	 * An object info holds a [.type] and a [.name].
	 * If the type is a Sprite it holds a list of frames. Otherwise it has a [.size] for debug drawing purposes.
	 * @author Trixt0r
	 */
	class ObjectInfo @JvmOverloads internal constructor(val name: String, val type: ObjectType, val size: Dimension, val frames: ArrayList<FileReference> = ArrayList<FileReference>()) {
		companion object {
			val DUMMY = ObjectInfo("", ObjectType.Bone, Dimension(0f, 0f))
		}

		internal constructor(name: String, type: ObjectType, frames: ArrayList<FileReference>) : this(name, type, Dimension(0f, 0f), frames) {}

		override fun toString(): String {
			return "$name: $type, size: $size|frames:\n$frames"
		}
	}

	/**
	 * Represents a Spriter SCML character map.
	 * A character map maps [FileReference]s to [FileReference]s.
	 * It holds an [CharacterMap.id] and a [CharacterMap.name].
	 * @author Trixt0r
	 */
	class CharacterMap(val id: Int, val name: String) : HashMap<FileReference, FileReference>() {
		companion object {
			val DUMMY = CharacterMap(0, "")
		}

		/**
		 * Returns the mapped reference for the given key.
		 * @param key the key of the reference
		 * *
		 * @return The mapped reference if the key is in this map, otherwise the given key itself is returned.
		 */
		override fun get(key: FileReference): FileReference? {
			if (!super.containsKey(key))
				return key
			else
				return super.get(key)
		}
	}

	override fun toString(): String {
		var toReturn = this::class.simpleName + "|[id: " + id + ", name: " + name + "]"
		toReturn += "Object infos:\n"
		for (info in this.objectInfos)
			toReturn += "\n" + info
		toReturn += "Character maps:\n"
		for (map in this.characterMaps)
			toReturn += "\n" + map
		toReturn += "Animations:\n"
		for (animaton in this.animations)
			toReturn += "\n" + animaton
		toReturn += "]"
		return toReturn
	}

	companion object {
		var DUMMY = Entity(0, "", 0, 0, 0)
	}

}
